### PR TITLE
Hide dropdown content when clicking away 

### PR DESCRIPTION
**Pull request by tylerbarna**
This adds basic functionality to the DetailsDropdownContent component for listening for clicking away from a dropdown, which will cause the dropdown to be hidden from the user unless the dropdown button is pressed again. This will prevent dropdown menus drawing over one another in the circulars archive like so:

image

and users won't have to manually hide dropdown menus should they choose to not impose any of the additional search filters.

### CODE REVIEW

**Comment by tylerbarna** [comment]
The current commit seems to interfere with the dropdown button interaction slightly; if the menu is expanded, pressing the button again does not collapse the menu, though clicking anywhere else on the page will collapse the dropdown. This is not the case in the currently deployed commit, so it seems to be due to the implementation made in this PR.

**Comment by tylerbarna** [comment]
@dakota002 would this affect the use of [DetailsDropdownContent in the schema browser docs](https://github.com/nasa-gcn/gcn.nasa.gov/blob/f657b161c35f3cd92ec2ecfb03f804c2be122b76/app/routes/_gcn.docs_._schema-browser.schema.(%24version).%24/route.tsx#L141)?

**Comment by dakota002** [comment]
> @dakota002 would this affect the use of [DetailsDropdownContent in the schema browser docs](https://github.com/nasa-gcn/gcn.nasa.gov/blob/f657b161c35f3cd92ec2ecfb03f804c2be122b76/app/routes/_gcn.docs_._schema-browser.schema.(%24version).%24/route.tsx#L141)?

Without checking, they should work well together, assuming that the property is not required. You should be able to test it out locally though. If you are having issues with the github content on your local let me know

**Review by lpsinger — CHANGES_REQUESTED** [review]
Please rebase.

Please use the existing hooks [useOnClickOutside](https://usehooks-ts.com/react-hook/use-on-click-outside) or [useClickAnywhere](https://usehooks-ts.com/react-hook/use-click-any-where).

**Comment by tylerbarna** [comment]
Change to useOnClickOutside changes issue from button not minimizing at all to it minimizing before immediately being expanded again; this is because the DetailsDropdownContent change now detects a click anywhere outside of the actual dropdown content, which then sets the expanded boolean to false, but clicking the button toggles the boolean, resulting in it being set to true again.

File: app/components/DetailsDropdownContent.tsx [hunk 1]
```diff
@@ -6,14 +6,25 @@
...
 export default function DetailsDropdownContent({
   children,
+  onClose,
   ...props
-}: Parameters<typeof Card>[0]) {
+}: Parameters<typeof Card>[0] & { onClose?: () => void }) {
+  const contentRef = useRef<HTMLDivElement>(null)
+
+  useOnClickOutside(contentRef, () => {
+    if (onClose) onClose()
```
Comment (lpsinger) [inline app/components/DetailsDropdownContent.tsx:20]:
```suggestion
    onClose?.()
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by tylerbarna** [comment]
Having some difficulty coming up with an adequately general way of detecting when a user is clicking an element we don't want to trigger the onClose with (eg detecting when the buttons in the buttongroup associated with the popup are clicked). Attempted to use both refs and classnames for this purpose with a check inside DetailsDropdownContent for one of those properties on the event.target, but there's variability in what the identified event.target is depending on where on the segmented button the user clicks

**Comment by tylerbarna** [comment]
this is not the most eloquent method, but the general idea is present. May be slightly better to use a ref rather than a classname, also some expression that checks all parents/grandparents/etc in the DetailsDropdownContent would be a bit better

**Comment by lpsinger** [comment]
Did you know that `useOnClickOutside` accepts a list of refs?

```ts
useOnClickOutside([dropdownContentRef, dropdownButtonRef], () => {...})`
```

**Comment by tylerbarna** [comment]
> Did you know that `useOnClickOutside` accepts a list of refs?
> 
> ```ts
> useOnClickOutside([dropdownContentRef, dropdownButtonRef], () => {...})
> ```

Looking back at our discussion on Monday, did you intend to have the useOnClickOutside defined outside of the DetailsDropdownContent component/have the refs defined in the route.tsx file instead of the DateSelectorMenu and SortSelectorMenu component files?

**Comment by lpsinger** [comment]
Yes, that would be the simplest.

File: app/routes/circulars._archive._index/SortSelectorButton.tsx [hunk 2]
```diff
@@ -51,63 +51,69 @@ function SortButton({
...
+  setShowContent: React.Dispatch<React.SetStateAction<boolean>>
+}
 
-  const submit = useSubmit()
+export const SortSelector = forwardRef<HTMLDivElement, SortSelectorProps>(
+  function SortSelector(
+    { form, defaultValue, showContent, setShowContent },
+    ref: React.Ref<HTMLDivElement>
+  ) {
+    // const [showContent, setShowContent] = useState(false)
```
Comment (Courey) [inline app/routes/circulars._archive._index/SortSelectorButton.tsx:66]: did you mean to delete this?

Comment (tylerbarna) [reply app/routes/circulars._archive._index/SortSelectorButton.tsx:66]: yes i did

File: app/routes/circulars._archive._index/DateSelectorMenu.tsx [hunk 3]
```diff
@@ -68,156 +68,168 @@ function DateSelectorButton({
...
+              <Grid col={4} key="radio-alltime">
                 <Radio
                   form=""
-                  id={`radio-${value}`}
+                  id="radio-alltime"
                   name="radio-date"
-                  value={value}
-                  label={label}
-                  defaultChecked={value === defaultStartDate}
+                  value=""
```
Comment (Courey) [inline app/routes/circulars._archive._index/DateSelectorMenu.tsx:154]: is this intentional?

Comment (Courey) [reply app/routes/circulars._archive._index/DateSelectorMenu.tsx:154]: oh I get it, nm.

File: app/routes/circulars._archive._index/route.tsx [hunk 4]
```diff
@@ -170,6 +171,18 @@ export default function () {
...
+  const [showDateSelector, setShowDateSelector] = useState(false)
+  const SortSelectorRef = useRef<HTMLDivElement>(null)
+  const [showSortSelector, setShowSortSelector] = useState(false)
+
+  useOnClickOutside(dateSelectorRef, () => {
+    setShowDateSelector(false)
+  })
+  useOnClickOutside(SortSelectorRef, () => {
+    setShowSortSelector(false)
+  })
```
Comment (lpsinger) [inline app/routes/circulars._archive._index/route.tsx:184]: These refs and states are only used inside `SortSelector` and `DateSelector`. Please move them inside those two respective components. They need not be exposed as props.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/circulars._archive._index/DateSelectorMenu.tsx [hunk 5]
```diff
@@ -77,7 +78,11 @@ export function DateSelector({
   defaultStartDate?: string
   defaultEndDate?: string
 }) {
-  const [showContent, setShowContent] = useState(false)
+  const dateSelectorRef = useRef<HTMLDivElement>(null)
+  const [showDateSelector, setShowDateSelector] = useState(false)
```
Comment (lpsinger) [inline app/routes/circulars._archive._index/DateSelectorMenu.tsx:82]: Renaming of the state from `showContent` to `showDateSelector` makes this a larger diff than necessary. Please keep the old name.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/circulars._archive._index/DateSelectorMenu.tsx [hunk 6]
```diff
@@ -77,7 +78,11 @@ export function DateSelector({
   defaultStartDate?: string
   defaultEndDate?: string
 }) {
+  const dateSelectorRef = useRef<HTMLDivElement>(null)
```
Comment (lpsinger) [inline app/routes/circulars._archive._index/DateSelectorMenu.tsx:81]: This is the only ref, so we can just call this variable `ref` without introducing ambiguity.

```suggestion
  const ref = useRef<HTMLDivElement>(null)
```

File: app/routes/circulars._archive._index/SortSelectorButton.tsx [hunk 7]
```diff
@@ -58,9 +59,12 @@ export function SortSelector({
   form?: string
   defaultValue?: string
 }) {
-  const [showContent, setShowContent] = useState(false)
-
   const submit = useSubmit()
+  const SortSelectorRef = useRef<HTMLDivElement>(null)
```
Comment (lpsinger) [inline app/routes/circulars._archive._index/SortSelectorButton.tsx:63]:
```suggestion
  const ref = useRef<HTMLDivElement>(null)
```

**Review by lpsinger — APPROVED** [review]

**Pull request merged**